### PR TITLE
Support custom [Packagist] server

### PR DESCRIPTION
**Pull request by honzahommer**
Add support for custom Packagist server instance.

### CODE REVIEW

**Comment by calebcartwright** [comment]
Thanks @honzahommer! A few questions for you:

1. Do you know of any publicly accessible custom Packagist servers that we could potentially use for the test targets?
2. I assume custom Packagist servers could require authentication? If so, we'd likely want to include support for that as part of this change, so that folks could use a self-hosted Shields instance to be able to interact with a private Packagist server.

**Comment by honzahommer** [comment]
@calebcartwright As mentioned in https://github.com/composer/packagist, Packagist is available under MIT Licence. This version has same behaviour as https://packagist.org/, co private packages are not supported. For private packages is Private Packagist (http://packagist.com), unfortunately I have no account there.

**Comment by calebcartwright** [comment]
So it sounds like the objective here is to support someone that is self-running their own public, internet accessible Packagist instance instead of using packagist.org. There is also self-hosted Private Packagist, but supporting that from Shields is not the objective of this PR.

Just out of curiosity, can you expand a bit on the use case? I'm trying to understand the circumstances in which someone would want to deploy their own self-hosted Packagist instance vs packagist.org, especially considering the below note from the documentation in the links you provided:

https://github.com/composer/packagist#packagist
> This project is not meant for re-use. It is open source to make it easy to contribute but we provide no support if you want to run your own, and will do breaking changes without notice

**Comment by honzahommer** [comment]
> So it sounds like the objective here is to support someone that is self-running their own public, internet accessible Packagist instance instead of using packagist.org. There is also self-hosted Private Packagist, but supporting that from Shields is not the objective of this PR.
> 
> Just out of curiosity, can you expand a bit on the use case? I'm trying to understand the circumstances in which someone would want to deploy their own self-hosted Packagist instance vs packagist.org, especially considering the below note from the documentation in the links you provided:
> 
> https://github.com/composer/packagist#packagist
> 
> > This project is not meant for re-use. It is open source to make it easy to contribute but we provide no support if you want to run your own, and will do breaking changes without notice

Yes, we are running self-hosted Packagist instance (it has some benefits from like high availability, hosting private repos, GitLab Oauth integration, etc.) integrated with GitLab (basically https://docs.gitlab.com/ee/user/project/integrations/project_services.html, but on steroids). So this is why we need change packagist server url for badge. I've just deployed branch [feat/packagist-custom-server](https://github.com/honzahommer/shields/tree/feat/packagist-custom-server) on our server, badges works fine.

**Comment by calebcartwright** [comment]
> hosting private repos,

That's the piece I was trying to zoom in on. Shields supports integration with private instances for various other services/platforms (see https://github.com/badges/shields/blob/master/doc/server-secrets.md#server-secrets for more info), and based on this thread, it seems like that may be applicable for our Packagist integration as well.

Ideally we'd also have that same type of support for our Packagist integration, however, if that's not of interest/in scope for you then I'd suggest we make a note of this in the documentation for the Packagist badges so that our end users will see that. I've made an inline code suggestion as a reference of this.

File: services/packagist/packagist-license.service.js [hunk 1]
```diff
@@ -34,6 +40,13 @@ module.exports = class PackagistLicense extends BasePackagistService {
         staticPreview: renderLicenseBadge({ license: 'MIT' }),
         keywords,
       },
+      {
+        title: 'Packagist (custom server)',
+        namedParams: { user: 'doctrine', repo: 'orm' },
+        queryParams: { server: 'https://packagist.org' },
+        staticPreview: renderLicenseBadge({ license: 'MIT' }),
+        keywords,
```
[the remark this answers is not in the document]
Comment (honzahommer) [reply services/packagist/packagist-license.service.js]: https://github.com/badges/shields/pull/3986/commits/fe65c565c443fe82c365e71227fdad55a88b0297

**Comment by honzahommer** [comment]
> Would you please also include the custom server support for the Packagist download badge too?

@calebcartwright Done (https://github.com/badges/shields/pull/3986/commits/180502327ef81540b4960ba60d57264e9b4e3d65).

**Comment by calebcartwright** [comment]
Looks like 3 of the tests for the downloads badges are failing

File: services/packagist/packagist-downloads.tester.js [hunk 2]
```diff
@@ -10,41 +10,117 @@ t.create('daily downloads (valid, no package version specified)')
...
+    message: isMetricOverTimePeriod,
+  })
+
+t.create(
+  'daily downloads (invalid, no package version specified, invalid custom server)'
+)
+  .get('/dd/doctrine/orm.json?server=https%3A%2F%2Fpackagist.com')
+  .expectBadge({
+    label: 'downloads',
+    message: 'badge not found',
```
Comment (calebcartwright) [inline services/packagist/packagist-downloads.tester.js]: IIRC this should be `inaccessible` for an invalid/unreachable host

**Pull request merged**